src/arkode/arkode_arkstep.c

```diff
@@ -2099,12 +2187,216 @@ int arkStep_TakeStep_Z(ARKodeMem ark_mem, sunrealtype* dsmPtr, int* nflagPtr)
   if (*nflagPtr < 0) { return (*nflagPtr); }
   if (*nflagPtr > 0) { return (TRY_AGAIN); }
 
+  if (ark_mem->checkpoint_scheme)
+  {
+    sunbooleantype do_save;
+    SUNAdjointCheckpointScheme_NeedsSaving(ark_mem->checkpoint_scheme,
+                                           ark_mem->checkpoint_step_idx,
+                                           step_mem->Be->stages,
+                                           ark_mem->tn + ark_mem->h, &do_save);
+    if (do_save)
+    {
+      SUNAdjointCheckpointScheme_InsertVector(ark_mem->checkpoint_scheme,
+                                              ark_mem->checkpoint_step_idx,
+                                              step_mem->Be->stages,
+                                              ark_mem->tn + ark_mem->h,
+                                              ark_mem->ycur);
+    }
+  }
+
   SUNLogExtraDebugVec(ARK_LOGGER, "updated solution", ark_mem->ycur, "ycur(:) =");
   SUNLogInfo(ARK_LOGGER, "end-compute-solution", "status = success");
 
   return (ARK_SUCCESS);
 }
 
+/*---------------------------------------------------------------
+  arkStep_TakeStep_ERK_Adjoint:
+
+  This routine performs a single backwards step of the discrete
+  adjoint of the ERK method.
+
+  Since we are not doing error control during the adjoint integration,
+  the output variable dsmPtr should should be 0.
+
+  The input/output variable nflagPtr is used to gauge convergence
+  of any algebraic solvers within the step. In this case, it should
+  always be 0 since we do not do any algebraic solves.
+
+  The return value from this routine is:
+            0 => step completed successfully
+           >0 => step encountered recoverable failure;
+                 reduce step and retry (if possible)
+           <0 => step encountered unrecoverable failure
+  ---------------------------------------------------------------*/
+int arkStep_TakeStep_ERK_Adjoint(ARKodeMem ark_mem, sunrealtype* dsmPtr,
+                                 int* nflagPtr)
+{
+  int retval = ARK_SUCCESS;
+
+  ARKodeARKStepMem step_mem;
+
+  /* access ARKodeARKStepMem structure */
+  retval = arkStep_AccessStepMem(ark_mem, __func__, &step_mem);
+  if (retval != ARK_SUCCESS) { return (retval); }
+
+  SUNLogDebug(ARK_LOGGER, "ARKODE::arkStep_TakeStep_ERK_Adjoint", "start-step",
+              "step = %li, h = %" SUN_FORMAT_G ", dsm = %" SUN_FORMAT_G
+              ", nflag = %d",
+              ark_mem->nst, ark_mem->h, *dsmPtr, *nflagPtr);
+
+  /* local shortcuts for readability */
+  SUNAdjointStepper adj_stepper = (SUNAdjointStepper)ark_mem->user_data;
```

Comment:
Maybe this will be clearer below, why attach the adjoint stepper as user data and not add it to `ARKodeMem`?

Reply:
We probably could add it to ARKodeMem, but also the user doesnt see this instance of ARKODE so hijacking the user_data is fine and doesnt add another pointer to ARKodMem.

Reply:
Actually, nevermind. It does need to be in user data because we need it in the RHS function.